File: py_solutions_21-30/Euler_28.py
# What is the sum of the numbers on the diagonals in a 1001 by 1001 spiral formed in the same way?

import timeit


start = timeit.default_timer()

def grid(step, axis):
    total = step
    for i in xrange(len(axis)):
        step += axis[i]
        if step > axis[len(axis) - 1]:
            break
        total += step
    return total

 
def euler_28():
    diag = 3
    total = sum(map(lambda x: x**2, xrange(1, 1002, 2)))
    for i in range(10, 15, 2):
        total += grid(diag, xrange(i, 1002002, 8))
        diag += 2
    return total


print "Answer: %s" % euler_28()
stop = timeit.default_timer()
print "Time: %f" % (stop - start)

File: Perl/perl_solutions_21-30/Euler_28.pl
use strict;
use warnings;
use Time::HiRes qw( clock );


sub grid
{
    my ($step, @axis) = @_;
    my $total = $step;
    for my $i (@axis) {
        $step += $i;
        if ($step > $axis[$#axis - 1]) {
            last;
        }
        $total += $step;
    }
    return $total;
}

sub spiral_path
{
    my $diag = 3;
    my $total = 0;
    my @path = (10, 12, 14);
    for my $i (1..1001) {
        if ($i % 2 != 0) {
            $total += $i**2;
        }
    }
    for my $i (@path) {
        my @spiral;
        do {
            push(@spiral, $i);
            $i += 8;
        } until ($i >= 1002002);
        $total += grid($diag, @spiral);
        $diag += 2; 
    }
    return $total;
}


my $start = clock();
my $answer = spiral_path();
my $stop = clock();

print "Answer: $answer\n";
printf "Time: %.5f\n", $stop - $start;
